parsec/backend/postgresql/utils.py

```diff
@@ -1,12 +1,233 @@
+        args = [self._stripped_sql]
+        for variable in self._variables:
+            args.append(kwargs[variable])
+        return args
+
+
+def q_organization(organization_id=None, _id=None, table="organization", select="*"):
+    assert organization_id is not None or _id is not None
+    if _id is not None:
+        condition = f"{table}._id = {_id}"
+    else:
+        condition = f"{table}.organization_id = {organization_id}"
+    return f"(SELECT {select} FROM {table} WHERE {condition})"
+
+
+def q_organization_internal_id(organization_id, **kwargs):
+    return q_organization(organization_id=organization_id, select="_id", **kwargs)
+
+
+def _table_q_factory(table, public_id_field):
+    def _q(
+        organization_id=None,
+        organization=None,
+        _id=None,
+        table_alias=None,
+        select="*",
+        suffix=None,
+        **kwargs,
+    ):
+        if table_alias:
+            from_table = f"{table} as {table_alias}"
+            select_table = table_alias
+        else:
+            from_table = table
+            select_table = table
+        if _id is not None:
+            condition = f"{select_table}._id = {_id}"
+        else:
+            public_id = kwargs.pop(public_id_field, None)
+            assert public_id is not None
+            assert organization_id is not None or organization is not None
+            if not organization:
+                organization = q_organization_internal_id(organization_id)
+            condition = f"{select_table}.organization = {organization} AND {select_table}.{ public_id_field } = { public_id }"
+        assert not kwargs
+        suffix = suffix or ""
+        return f"(SELECT {select} FROM {from_table} WHERE {condition} {suffix})"
+
+    def _q_internal_id(**kwargs):
+        return _q(select="_id", **kwargs)
+
+    _q.__name__ = f"q_{table}"
+    _q_internal_id.__name__ = f"q_{table}_internal_id"
+
+    return _q, _q_internal_id
+
+
+q_device, q_device_internal_id = _table_q_factory("device", "device_id")
+q_user, q_user_internal_id = _table_q_factory("user_", "user_id")
+q_realm, q_realm_internal_id = _table_q_factory("realm", "realm_id")
+q_block, q_block_internal_id = _table_q_factory("block", "block_id")
+q_human, q_human_internal_id = _table_q_factory("human", "email")
+
+
+def q_vlob_encryption_revision_internal_id(
+    encryption_revision,
+    organization_id=None,
+    organization=None,
+    realm_id=None,
+    realm=None,
+    table="vlob_encryption_revision",
+):
+    if realm is None:
+        assert realm_id is not None
+        assert organization_id is not None or organization is not None
+        if organization is None:
+            _q_realm = q_realm_internal_id(organization_id=organization_id, realm_id=realm_id)
+        else:
+            _q_realm = q_realm_internal_id(organization=organization, realm_id=realm_id)
+    else:
+        _q_realm = realm
+    return f"""
+(SELECT _id
```

Comment:
```
(
SELECT _id
FROM {table}
WHERE
    {table}.realm = {_q_realm}
    AND {table}.encryption_revision = {encryption_revision}
)
```
I think it makes the parenthesis more visible (and vertical indentation between SELECT and FROM is not consistent)